federation/pkg/federation-controller/util/federated_updater.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"time"
+
+	federation_release_1_4 "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	pkg_runtime "k8s.io/kubernetes/pkg/runtime"
+)
+
+// Type of the operation that can be executed in Federated.
+type FederatedOperationType string
+
+const (
+	OperationTypeAdd    = "add"
+	OperationTypeUpdate = "update"
+	OperationTypeDelete = "delete"
+)
+
+// FederatedOperation definition contains type (add/update/delete) and the object itself.
+type FederatedOperation struct {
+	Type FederatedOperationType
+	Obj  pkg_runtime.Object
+}
+
+// A helper that executes the given set of updates on federation, in parallel.
+type FederatedUpdater interface {
+	// Executes the given set of operations within the specified timeout.
+	// Timeout is best-effort. There is no guarantee that the underlying operations are
+	// stopped when it is reached. However the function will return after the timeout
+	// with a non-nil error.
+	Update([]FederatedOperation, time.Duration) error
+}
+
+// A function that executes some operation using the passed client and object.
+type FederatedOperationHandler func(federation_release_1_4.Interface, pkg_runtime.Object) error
+
+type federatedUpdaterImpl struct {
+	federation FederationView
+
+	addFunction    FederatedOperationHandler
+	updateFunction FederatedOperationHandler
+	deleteFunction FederatedOperationHandler
+}
+
+func NewFederatedUpdater(federation FederationView, add, update, del FederatedOperationHandler) FederatedUpdater {
+	return &federatedUpdaterImpl{
+		federation:     federation,
+		addFunction:    add,
+		updateFunction: update,
+		deleteFunction: del,
+	}
+}
+
+func (fu *federatedUpdaterImpl) Update(ops []FederatedOperation, timeout time.Duration) error {
+	done := make(chan error, len(ops))
+	for _, op := range ops {
+		go func(op FederatedOperation) {
+			clusterName, err := GetClusterName(op.Obj)
```

Comment:
it has to be the apiserver as there will be other clients access to it via rest api directly.
two reasons I prefer not use the cluster reference here
1. cluster reference may have more field then this code will need to update to populate the new fields
2. this will cause the cluster reference saved to etcd on early versions.
On Aug 12, 2016 5:45 PM, Marcin Wielgus <notifications@github.com> wrote:In federation/pkg/federation-controller/util/federated_updater.go:
> +}
> +
> +func NewFederatedUpdater(federation FederationView, add, update, del FederatedOperationHandler) FederatedUpdater {
> +	return &federatedUpdaterImpl{
> +		federation:     federation,
> +		addFunction:    add,
> +		updateFunction: update,
> +		deleteFunction: del,
> +	}
> +}
> +
> +func (fu *federatedUpdaterImpl) Update(ops []FederatedOperation, timeout time.Duration) error {
> +	done := make(chan error, len(ops))
> +	for _, op := range ops {
> +		go func(op FederatedOperation) {
> +			clusterName, err := GetClusterName(op.Obj)

I think it will be also handy for federated clientset to actually populate/remove this as well. I see little point of having a field that is populated only on special occasions. Sure, we can add this information as a separate field in federatedoperation and elsewhere but why? 

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.